elegant/edge_detection.py

```diff
@@ -0,0 +1,203 @@
+    (NOTE: this is the way that elegant generates the straightened worm panel in the gui)
+
+    Parameters:
+        image: ndarray of the straightened worm image
+        avg_width_tck: width spline defining the average distance from the centerline
+            to the worm edges (This is taken from the pca things we did earlier)
+
+    Returns:
+        center_coordinates: shape (num_coords, 2) list of coordinates that define the centerline
+            in the worm frame of reference
+        width_coordinates: shape (num_coords, 2) list of coordinates that define the widths
+            in the worm fram of reference
+    """
+    #break up the image into top and bottom parts to find the widths on each side
+    top_image = np.flip(image[:,:int(image.shape[1]/2)], axis =1)
+    bottom_image = image[:,int(image.shape[1]/2):]
+    xt, top_widths = find_edges(top_image, avg_width_tck)
+    xb, bottom_widths = find_edges(bottom_image, avg_width_tck)
+
+    top_widths = top_widths.astype(np.float)
+    bottom_widths = bottom_widths.astype(np.float)
+
+    #need to account for the top and bottom splitting sincethe centerline is exactly 
+    #in the middle of the pixels, we need to add 0.5 to the widths
+    if (image.shape[1]/2) % 2 == 0:
+        top_widths += 0.5
+        bottom_widths += 0.5
+
+    #NOTE: using the midpoint as the zero axis, we negate the bottom widths (hence, it is
+    #negative here) ie. it becomes (top_widths + (-bottom_widths))
+    new_center = (bottom_widths-top_widths)/2 #find the midpoint between the two points
+    #put the pixels in the same frame of reference as straightened worm
+    new_center += int(image.shape[1]/2)
+    new_widths = (top_widths+bottom_widths)/2#calculate the average widths
+
+    center_coordinates = np.transpose([xt,new_center])
+    width_coordinates = np.transpose([xt,new_widths])
+
+    return (center_coordinates, width_coordinates)
+
+
+def find_edges(image, avg_width_tck, ggm_sigma=1, sig_per=61, sig_growth_rate=2, alpha=1, mcp_alpha=1):
+    """Find the edges of one side of the worm and return the x,y positions of the new widths
+    NOTE: This function assumes that the image is only half of the worm (ie. from the centerline
+    to the edges of the worm)
+
+    Parameters:
+        image: ndarray of the straightened worm image (typically either top or bottom half)
+        avg_width_tck: width spline defining the average distance from the centerline
+            to the worm edges (This is taken from the pca things we did earlier)
+        ggm_sigma, sig_per, sig_growth_rate, alpha, mcp_alpha: hyperparameters for 
+            the edge-detection scheme
+    
+    Returns:
+        route: tuple of x,y positions of the identfied edges
+    """
+
+    #down sample the image
+    image_down = pyramid.pyr_down(image, downscale=2)
+
+    #get the gradient
+    gradient = ndimage.filters.gaussian_gradient_magnitude(image_down, 1)
```

Comment:
you're not actually using the `ggm_sigma`, `sig_per`, or `sig_growth_rate` parameters here! Hopefully the code you were optimizing on is not this version with the parameters hard-coded... 

Probably should also describe those parameters a little more clearly in the documentation.

Also, except for the `sig_growth_rate` parameter, all of this stuff is completely independent of the absolute image intensities (because it's defined in terms of percentiles). So the `scale_image` stuff is almost dispensable. I wonder if there's a way to represent the growth rate in terms of the gradient min and max or similar?